### ISSUE TITLE
Distributed Training & Quantization Techniques

### ISSUE THREAD

**Issue by sujith2303** [sujith2303/ML_From_Scratch#8]
- [ ] FSDP
- [ ] ZERO
- [ ] QLoRA
- [ ] GRPO
- [ ] Bits & Bytes GPTQ-Int8,GPTQ-Int4, AWQ
- [ ] Learn how to deal with data , how to load data in different ways
- [x] Learn different losses like Cross Entropy, KL, etc and implement from scratch.
- [ ] Learn about optimizers and implement the Adam optimizer from scratch similar to the Pytorch's loss.backward() and optimizers.step()
- [ ]  RoPE, SwiGLU, RMSNorm, Attention QKV bias and tied word embeddings
- [ ] Tokenization from scratch
